mlir/test/Quantum/MergeRotationsTest.mlir

```diff
@@ -0,0 +1,255 @@
+    // CHECK-NOT: quantum.custom "CRX"
+    %3:2 = quantum.custom "ControlledPhaseShift"(%arg0) %1, %2: !quantum.bit, !quantum.bit
+    %4:2 = quantum.custom "ControlledPhaseShift"(%arg1) %3#0, %3#1 : !quantum.bit, !quantum.bit
+    %5:2 = quantum.custom "ControlledPhaseShift"(%arg2) %4#0, %4#1 : !quantum.bit, !quantum.bit
+    // CHECK: return [[ret]]#0, [[ret]]#1
+    return %5#0, %5#1 : !quantum.bit, !quantum.bit
+}
+
+// -----
+
+func.func @test_merge_rotations(%arg0: f64, %arg1: f64, %arg2: f64) -> (!quantum.bit, !quantum.bit) {
+    %0 = quantum.alloc( 2) : !quantum.reg
+    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
+    %2 = quantum.extract %0[ 1] : !quantum.reg -> !quantum.bit
+
+    // CHECK: [[reg:%.+]] = quantum.alloc( 2) : !quantum.reg
+    // CHECK: [[qubit1:%.+]] = quantum.extract [[reg]][ 0] : !quantum.reg -> !quantum.bit
+    // CHECK: [[qubit2:%.+]] = quantum.extract [[reg]][ 1] : !quantum.reg -> !quantum.bit
+    // CHECK: [[sum:%.+]] = arith.addf %arg0, %arg1 : f64
+    // CHECK: [[qubits3:%.+]]:2 = quantum.custom "ControlledPhaseShift"([[sum]]) [[qubit1]], [[qubit2]] : !quantum.bit, !quantum.bit
+    // CHECK: [[ret:%.+]]:2 = quantum.custom "ControlledPhaseShift"(%arg2) [[qubits3]]#1, [[qubits3]]#0 : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.custom "ControlledPhaseShift"
+    %3:2 = quantum.custom "ControlledPhaseShift"(%arg0) %1, %2: !quantum.bit, !quantum.bit
+    %4:2 = quantum.custom "ControlledPhaseShift"(%arg1) %3#0, %3#1 : !quantum.bit, !quantum.bit
+    %5:2 = quantum.custom "ControlledPhaseShift"(%arg2) %4#1, %4#0 : !quantum.bit, !quantum.bit
+    // CHECK: return [[ret]]#0, [[ret]]#1
+    return %5#0, %5#1 : !quantum.bit, !quantum.bit
+}
+
+// -----
+
+func.func @test_merge_rotations(%arg0: f64, %arg1: f64, %arg2: f64) -> !quantum.bit {
+    // CHECK: [[reg:%.+]] = quantum.alloc( 1) : !quantum.reg
+    // CHECK: [[qubit:%.+]] = quantum.extract [[reg]][ 0] : !quantum.reg -> !quantum.bit
+    %0 = quantum.alloc( 1) : !quantum.reg
+    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
+    // CHECK:    [[angle00:%.+]] = arith.addf %arg1, %arg2 : f64
+    // CHECK:    [[angle10:%.+]] = arith.addf %arg2, %arg0 : f64
+    // CHECK:    [[angle20:%.+]] = arith.addf %arg0, %arg1 : f64
+    // CHECK:    [[angle01:%.+]] = arith.addf %arg0, [[angle00]] : f64
+    // CHECK:    [[angle11:%.+]] = arith.addf %arg1, [[angle10]] : f64
+    // CHECK:    [[angle21:%.+]] = arith.addf %arg2, [[angle20]] : f64
+    // CHECK: [[ret:%.+]] = quantum.custom "Rot"([[angle01]], [[angle11]], [[angle21]]) [[qubit]] : !quantum.bit
+    // CHECK-NOT: quantum.custom "Rot"
+    %2 = quantum.custom "Rot"(%arg0, %arg1, %arg2) %1 : !quantum.bit
+    %3 = quantum.custom "Rot"(%arg1, %arg2, %arg0) %2 : !quantum.bit
+    %4 = quantum.custom "Rot"(%arg2, %arg0, %arg1) %3 : !quantum.bit
+    // CHECK: return [[ret]]
+    return %4 : !quantum.bit
+}
+
+
```

Comment:
Can we add a test for explicit control wires? e.g. https://github.com/PennyLaneAI/catalyst/blob/main/mlir/test/Quantum/ChainedSelfInverseTest.mlir#L399

I know there's qml.crot but explicitly controlled rotation gates might also exist from qml.ctrl, for example `qml.ctrl(some_function)` and the function has rotation gates, and the explicit control wires got distributed onto the rotation gates

Reply:
thanks for the link I will add that

Reply:
added